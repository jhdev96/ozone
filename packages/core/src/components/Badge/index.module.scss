@import '../../scss/mixins/badge/badge.scss';


$positionOffset: 7px;

.badge {
  @include badge;
}

/* Main badge classes */
.primary {
  color: #fff;
  background-color: var(--blue-core);
}

.success {
  color: var(--green-3);
  background-color: var(--green-1);
}

.info {
  color: var(--blue-7);
  background-color: var(--blue-2);
}

.warning {
  color: var(--yellow-3);
  background-color: var(--yellow-1);
}

.secondary {
  color: #fff;
  background-color: var(--red-core);
}

.routine {
  color: var(--black-2);
  background-color: var(--gray-core);
}

.rounded { 
  padding: 4px;
  min-width: fit-content;
  width: 12.95px;
  height: 12px;
  border-radius: 50%; 
}

/* Positioning */
.absolute { 
  position: absolute !important; 
  margin: 0 !important;
}

.topLeft { 
  top: -$positionOffset; 
  left: -$positionOffset;
}

.topRight {
  top: -$positionOffset;
  right: -$positionOffset;
}

.bottomLeft {
  bottom: -$positionOffset;
  left: -$positionOffset;
}

.bottomRight {
  bottom: -$positionOffset;
  right: -$positionOffset;
}