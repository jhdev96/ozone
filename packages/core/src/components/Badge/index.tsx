import React from 'react';
import classNames from '../../utils/classNames';
import styles from './index.module.scss';


export interface BadgeProps {
  /** 
   * Main classes for a badge. 
   */
  variation?: "primary"   | 
              "secondary" | 
              "info"      |
              "success"   | 
              "warning"   | 
              "routine";
  /** 
   * Determine if the shape of the badge is circular. 
   */
  circular?: boolean;
  /** 
   * When true, the badge will be given a css
   * class with a `position: absolute` rule.
   */ 
  absolute?: boolean;
  /** 
   * Provide additional css styling to the badge.
   */ 
  className?: string;
  /** 
   * Determines where the badge is placed relative 
   * to the target element.
   * 
   * This only takes effect when the `position`
   * prop is set to `absolute`.
   */ 
  placement?: "topLeft"     | 
              "topRight"    | 
              "bottomLeft"  | 
              "bottomRight";
  /** 
   * Aria label for screen readers. 
   */ 
  ariaLabel?: string;
  /**
   * A selector used in test environments.
   */
  dataTestId?: string;
  /** 
   * The content rendered inside a badge. 
   */
  children?: React.ReactNode;
};

export default function Badge({
  variation = "routine",
  circular  = false,
  absolute,
  className,
  placement = "topRight",
  ariaLabel,
  dataTestId,
  children
}: BadgeProps): JSX.Element 
{
  return (
    <span 
      className={classNames(
        className,
        styles.badge, 
        styles[variation],
        absolute    && styles.absolute,
        circular    && styles.rounded,
        absolute    && styles[placement]
      )}
      aria-label={ariaLabel}
      data-testid={dataTestId}
    >
      {children}
    </span>
  )
}

Badge.displayName = 'Badge';