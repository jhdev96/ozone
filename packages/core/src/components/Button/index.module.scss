@import '../../scss/root';
@import '../../scss/mixins/button/button.scss';
@import '../../scss/mixins/button/variables.scss';
                                  

// For styling `flat` buttons
$flatMap: (
  primary: (
    color: #fff,
    bg-color: var(--blue-3)
  ),
  secondary: (
    color: #fff,
    bg-color: var(--red-3)
  ),
  warning: (
    color: var(--black-core),
    bg-color: var(--yellow-3)
  )
);

/* Major button classes - primary, secondary, warning, basic */                                   
.primary {
  @include button;

  color: #fff;
  background-color: var(--blue-core);
  border-radius: 4px;

  &--outline {
    @extend .flat;

    color: var(--blue-core);
    border: 1px solid var(--blue-core) !important;
    background-color: transparent;
  }

  &:focus {
    outline: var(--red-core) auto 1px !important;
  }
}

.secondary {
  @include button;

  color: #fff;
  background-image: $secondary-gradient;
  background-repeat: repeat-x;
  border-color: $button-bc1 $button-bc2 $button-bc3 !important;
  box-shadow: inset 0 1px 0 $primary-shade1, 
                    0 1px 2px $primary-shade2;
  background-color: var(--red-core);
  border-radius: 4px;

  &--outline {
    @extend .flat;
    color: var(--red-core);
    border: 1px solid var(--red-core) !important;
    background-color: transparent;
  }

  &:focus {
    outline: var(--blue-core) auto 1px !important;
  }
}

.warning {
  @include button;

  color: var(--black-core);
  background-color: var(--yellow-core);
  border-radius: 4px;

  &:focus {
    outline: var(--black-core) auto 1px !important;
  }
}

.basic {
  @include button;

  border-radius: 3px;

  &:focus {
    outline: var(--gray-4) auto 1px !important;
  }
}

/* The following styles serve as modifiers 
 * for the major button classes 
*/
.flat {
  background-image: none !important;
  background-repeat: no-repeat;
  border: 0;
  border-radius: 0 !important;
  box-shadow: none !important;
  text-shadow: none !important;

  @each $class, $rule in $flatMap {
    &.#{$class}:not(.disabled):not(.loading) {
      &:hover {
        color: map-get($rule, 'color');
        background-color: map-get($rule, 'bg-color');
      }
    }
  }
}

.fill { width: 100%; }

.loading {
  padding: 8px 20px 5px 20px;
  max-height: 35px;
  opacity: 0.90;
  cursor: default;

  &:focus {outline: none !important;}
}

.disabled {
  opacity: 0.70;
  cursor: default;
}

.small {
  min-width: fit-content;
  padding: 8px !important;
}

.medium { min-width: 40px; }

.large {
  min-width: 70px;
  max-height: 70px !important;
  padding: 15px 20px !important;
  font-size: 1.2em !important;
}

.light {
  &.basic {
    color: var(--black-core);
    background-color: #fff;
    border: 1px solid var(--gray-3);
  }
}

.dark {
  &.basic {
    color: var(--gray-1);
    border: 1px solid var(--black-6);
    background-color: var(--black-4) !important;
  }
}