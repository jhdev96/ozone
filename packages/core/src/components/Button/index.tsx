import React from 'react';
import classNames from '../../utils/classNames';
import { 
  ButtonVariation,
  loaderTheme
} from './utils';
import LoadingDots from '../LoadingDots';
import styles from './index.module.scss';


export interface ButtonProps {
  /** 
   * The main classes of buttons. 
   */
  variation?: ButtonVariation;
  /** 
   * Sets the HTML `type` attribute for the
   * button element.
   */ 
  type?: React.ButtonHTMLAttributes<HTMLButtonElement>["type"]
  /** 
   * Control the size of the button. Default is medium. 
   */
  size?: "small" | "medium" | "large";
  /** 
   * Give the button a flat texture, appropriate for forms. 
   */
  flat?: boolean;
  /** 
   * When true the button will no background; 
   * It will only have a border. 
   */
  outline?: boolean;
  /** 
   * When true, the button's width will be 100% of 
   * its container. 
   */ 
  fill?: boolean;
  /** 
   * Color the button based on the current theme. 
   */
  theme?: "light" | "dark";
  /** 
   * Make the button disabled. 
   */
  isDisabled?: boolean;
  /** 
   * Indicates whether the button is loading some content. 
   */
  isLoading?: boolean;
  /** 
   * Sets the `className` attribute on the button element. 
   */ 
  className?: string;
  /** 
   * The function that handles click events. 
   */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Called when the user's mouse hovers over the button.
   */
  onMouseEnter?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Called when the user's mouse hovers away from the
   * button.
   */
  onMouseLeave?: React.MouseEventHandler<HTMLButtonElement>;
  /** 
   * A selector used for testing this component. 
   */
  dataTestId?: string;
  /** 
   * Content displayed inside the button. 
   */
  children?: React.ReactNode | string;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  function Button({
    variation     = "primary",
    type          = "button",
    size          = "medium", 
    flat          = false,
    outline       = false,
    fill          = false, 
    theme         = "light",
    isDisabled    = false,
    isLoading     = false,
    className,
    onClick       = () => {},
    onMouseEnter  = () => {},
    onMouseLeave  = () => {},
    dataTestId,
    children}: ButtonProps,
    ref: React.Ref<HTMLButtonElement>): JSX.Element 
  {
    const childMarkup = isLoading 
      ? (
          <>
            <LoadingDots 
              dotSize={7} 
              position="center" 
              theme={loaderTheme(variation, theme, outline)} 
            />
            <span style={{opacity: '0'}}>{children}</span>
          </>
        )
      : children;

    return (
      <button
        type={type}
        className={classNames(
          className,
          styles[variation],
          styles[size],
          styles[theme],
          flat        && styles.flat,
          outline     && styles[`${variation}--outline`],
          fill        && styles.fill,
          isLoading   && styles.loading,
          isDisabled  && styles.disabled
        )}
        onClick={isDisabled ? () => {} : onClick}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        disabled={isDisabled}
        aria-busy={isLoading}
        data-testid={dataTestId}
        ref={ref}
      >
        {childMarkup}
      </button>
    );
  }
);

Button.displayName = 'Button';

export default Button;