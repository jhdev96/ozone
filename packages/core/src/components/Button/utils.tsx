import { LoadingDotsProps } from '../LoadingDots';


export type ButtonVariation = "primary" | "secondary" | "warning" | "basic";

export const loaderTheme = (
  btnType: ButtonVariation, 
  theme: "light" | "dark", 
  outline: boolean
): LoadingDotsProps["theme"] => {
  /** 
   * Sets the theme for the LoadingDots component
   * which is rendered as a child for a button
   * that is loading. 
   * 
   * @param btnType // the type of button wrapping the loader
   * @param theme // the button's theme
   * @param outline // whether the button has just an outline
   * 
   * This function will return
   * an appropriate theme for the LoadingDots
   * component, so that it contrasts with the
   * button.
   */
  switch(btnType) {
    case 'basic':
      if (theme === 'light') return 'dark';
      return 'light';
    case 'primary':
      if (outline) return "blue";
      return theme;
    case 'secondary':
      if (outline) return "red";
      return theme;
    case 'warning':
      return "dark";
    default:
      return theme;
  }
}