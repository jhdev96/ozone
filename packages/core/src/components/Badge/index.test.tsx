import React from 'react';
import { render, screen } from '@testing-library/react';
import Button from '../Button';
import Badge, { BadgeProps } from '.';


describe('<Badge />', () => {
  function renderBadgeWithPlacement(
    placement: BadgeProps["placement"], 
    testId: string
  ) 
  {
    render(
      <Button>
        Inbox
        <Badge 
          placement={placement} 
          dataTestId={testId}
          absolute
        >
          10
        </Badge>
      </Button>
    );
  }

  it('renders a badge with children', () => {
    const text = "This is some text";
    render(
      <Badge dataTestId="badge">{text}</Badge>
    );

    expect(screen.getByTestId('badge'))
      .toHaveTextContent(text)
  });

  it('renders a circular badge', () => {
    render(
      <Badge dataTestId="badge" circular>1</Badge>
    );

    expect(screen.getByTestId('badge'))
      .toMatchSnapshot();
  });

  it('renders a badge with the correct placement', () => {
    renderBadgeWithPlacement('topLeft', 'test1');
    renderBadgeWithPlacement('topRight', 'test2');
    renderBadgeWithPlacement('bottomLeft', 'test3');
    renderBadgeWithPlacement('bottomRight', 'test4');

    expect(screen.getByTestId('test1')).toMatchSnapshot('topLeft');
    expect(screen.getByTestId('test2')).toMatchSnapshot('topRight');
    expect(screen.getByTestId('test3')).toMatchSnapshot('bottomLeft');
    expect(screen.getByTestId('test4')).toMatchSnapshot('bottomRight');
  });

  it('renders a badge with an `aria-label` attribute', () => {
    render(
      <Badge 
        variation="primary" 
        ariaLabel="unread messages"
        dataTestId="badge"
      >
        200
      </Badge>
    );

    expect(screen.getByTestId('badge').getAttribute('aria-label'))
      .toEqual('unread messages');
  });
});