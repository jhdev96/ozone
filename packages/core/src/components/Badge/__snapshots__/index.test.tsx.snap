// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge /> renders a badge with the correct placement: bottomLeft 1`] = `
<span
  class="badge routine absolute bottomLeft"
  data-testid="test3"
>
  10
</span>
`;

exports[`<Badge /> renders a badge with the correct placement: bottomRight 1`] = `
<span
  class="badge routine absolute bottomRight"
  data-testid="test4"
>
  10
</span>
`;

exports[`<Badge /> renders a badge with the correct placement: topLeft 1`] = `
<span
  class="badge routine absolute topLeft"
  data-testid="test1"
>
  10
</span>
`;

exports[`<Badge /> renders a badge with the correct placement: topRight 1`] = `
<span
  class="badge routine absolute topRight"
  data-testid="test2"
>
  10
</span>
`;

exports[`<Badge /> renders a circular badge 1`] = `
<span
  class="badge routine rounded"
  data-testid="badge"
>
  1
</span>
`;
